sql: add catalog.Column interface, replace catalog.TableDescriptor methods for columns

postamar opened the pull request:
In an effort to not directly use `descpb.ColumnDescriptor` protos everywhere, this multi-commit PR:
1. introduces a new `catalog.Column` interface to encapsulate them, and a bunch of new methods in `catalog.TableDescriptor` to replace those which have `descpb.ColumnDescriptor` in their signatures, this is done in the first commit;
2. replaces calls to the old methods with calls to the new methods throughout the codebase in the subsequent commits.

This breakdown into multiple commits is done to ease review, most of the substantial changes are in the first commit, the others follow through with the change and so are quite noisy.

Fixes #59805. 

See also the parent issue #56306, as well as the related issue #57465 which tracks the same work I've already done for `descpb.IndexDescriptor`. In this PR I reused many of the same patterns I introduced in the corresponding PR #58678.

Subsequent work should consist in propagating this `descpb.ColumnDescriptor -> catalog.Column` change further throughout the codebase.

Commit:
sql/catalog: add Column interface and TableDescriptor methods

Previously, the catalog.TableDescriptor interface and its implementing
types would liberally return descpb.ColumnDescriptor values, pointers and
slices. In an effort to stop manipulating such protos directly, this
patch introduces a catalog.Column interface to encapsulate it. In order
to enventually propagate this change throughout the code base, this
patch marks some existing catalog.TableDescriptor methods as deprecated
and introduces new ones to eventually replace them.

This work is very similar to that which I already performed for #57465.

Fixes #59805.

Release note: None

Commit:
sql: remove deprecated FindColumn-methods from TableDescriptor

Previously, this method would be called on a table descriptor interface to
return a target *descpb.ColumDescriptor.

This patch removes these calls, along with the method definitions, in
favour of new methods and functions which use the catalog.Column interface
type instead.

Fixes #59805.

Release note: None

Commit:
sql: remove deprecated column slice methods in TableDescriptor

Previously, these methods would be called on a table descriptor interface
to obtain []descpb.ColumnDescriptor values.

This patch removes these calls, along with the method definitions, in
favour of new methods which use the catalog.Column interface type instead.

GetPublicColumns() is treated in a separate commit owing to its large
number of call sites.

Fixes #59805.

Release note: None

Commit:
sql: remove deprecated ForeachColumn-methods from TableDescriptor

Previously, these methods would be called on a table descriptor interface
to apply a function on *descpb.ColumnDescriptor for some subset of columns.

This patch removes these calls, along with the method definitions, in favour
of new methods which use the catalog.Column interface type instead.

Fixes #59805.

Release note: None

Commit:
sql: remove deprecated ColumnIdx-methods from TableDescriptor

Previously, the ColumnIdxMap* methods and the GetColumnAtIdx method would be
called on a table descriptor interface to return catalog.TableColMap and
corresponding descpb.ColumnDescriptor values, respectively.

This patch removes these calls, along with the method definitions, in
favour of new methods and functions which use the catalog.Column interface
type instead.

Fixes #59805.

Release note: None

Commit:
sql: remove deprecated GetPublicColumns method from TableDescriptor

Previously, this method would be called on a table descriptor interface
to return a []descpb.ColumnDescriptor value containing all public
columns.

This patch removes these calls, along with the method definition, in
favour of new methods which use the catalog.Column interface type instead.

Fixes #59805.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59789)
<!-- Reviewable:end -->

Commit:
sql: rename Column slice methods in TableDescriptor

This patch renames the catalog.Column-slice-returning methods which I
recently added to catalog.TableDescriptor. I gave them provisional names
to differentiate them from the existing and very similar deprecated methods
that they replaced. Now that all deprecated methods have been removed we
can give them their definitive names.

Fixes #59805.

Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> just tiny comment things. 

> Reviewed 10 of 10 files at r1, 10 of 10 files at r2, 36 of 36 files at r3, 22 of 22 files at r4, 25 of 25 files at r5, 48 of 48 files at r6, 70 of 70 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-:-MSkNx_Q3Wfw4Lf_ZN4z:bmef8fr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru and @postamar)

___
*[pkg/sql/opt_catalog.go, line 638 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-MSk3R2b-ZmNTIUgb28L:-MSk3R2b-ZmNTIUgb28M:b-iy351p) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf1d1ad163892ea0ddb82a7b6ff533742fb7e34c/pkg/sql/opt_catalog.go#L638)):*
> ```Go
> 	}
> 
> 	ot.columns = make([]cat.Column, len(cols), numCols)
> ```

this `cat.Column` thing is something to pay attention to. 
___
*[pkg/sql/opt_exec_factory.go, line 1956 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-MSkMkehC5Q2beLbip55:-MSkMkei3nGFBHxrPlZ-:b-5pp50v) ([raw file](https://github.com/cockroachdb/cockroach/blob/52705f226fadfcc26b8a534ed0852a9d28fc9935/pkg/sql/opt_exec_factory.go#L1956)):*
> ```Go
> }
> 
> func makePublicColIdxToRetIdx(
> ```

The name of this function doesn't clearly describe to me its behavior; a comment would be helpful. 
___
*[pkg/sql/catalog/descriptor.go, line 404 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-MSiSVCX887V8-bbctkW:-MSiSVCY4cyrRs8BkzWt:bh37071) ([raw file](https://github.com/cockroachdb/cockroach/blob/50373fa7baf8df4655a6e7780952811946eb61d9/pkg/sql/catalog/descriptor.go#L404)):*
> ```Go
> 	// The ordinal of a column in a `tableDesc descpb.TableDescriptor` is
> 	// defined as follows:
> 	// - [0:len(tableDesc.Columns)] is the range of public columns,
> ```

Should we comment this in terms of methods rather than fields?
___
*[pkg/sql/catalog/descriptor.go, line 411 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-MSiSi7OCuUVf-uNO64y:-MSiSi7OCuUVf-uNO64z:bz3lpe8) ([raw file](https://github.com/cockroachdb/cockroach/blob/50373fa7baf8df4655a6e7780952811946eb61d9/pkg/sql/catalog/descriptor.go#L411)):*
<details><summary><i alt="summary">Quoted 15 lines of code…</i></summary>

> ```
> 	// Public returns true iff the column is active, i.e. readable, in the table
> 	// descriptor.
> 	Public() bool
> 	// WriteAndDeleteOnly returns true iff the column is a mutation in the
> 	// delete-and-write-only state in the table descriptor.
> 	WriteAndDeleteOnly() bool
> 	// DeleteOnly returns true iff the column is a mutation in the delete-only
> 	// state in the table descriptor.
> 	DeleteOnly() bool
> 	// Adding returns true iff the column is an add mutation in the table
> 	// descriptor.
> 	Adding() bool
> 	// Dropped returns true iff the column is a drop mutation in the table
> 	// descriptor.
> 	Dropped() bool
> ```
</details>

Are we happy with these as booleans? I know we did this on indexes but I have mixed feelings about it vs. an enum. 

----

After having reviewed this a bit, I think these works out pretty well, :shrug: 
___
*[pkg/sql/catalog/tabledesc/table.go, line 466 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-MSk-8lA1vxJEfVDQ01U:-MSk-8lA1vxJEfVDQ01V:b-lx5yt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/50373fa7baf8df4655a6e7780952811946eb61d9/pkg/sql/catalog/tabledesc/table.go#L466)):*
> ```Go
> // like FindPublicColumnWithName applied repeatedly to the names in the
> // provided list.
> func FindPublicColumnsWithNames(
> ```

Note that an error will be returned if any of the names do not exist. 


<!-- Sent from Reviewable.io -->

postamar reviewed (commented):
Thanks for reviewing and for the quick turnaround. I'm going to hurry up and merge this before more conflicts arise. I already had to rebase once. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-:-MSmYXR38l_e9l2BnMsX:b5058yo)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru and @ajwerner)

___
*[pkg/sql/opt_catalog.go, line 638 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-MSk3R2b-ZmNTIUgb28L:-MSmYkq8B68vno0irC3B:b-cznjps) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf1d1ad163892ea0ddb82a7b6ff533742fb7e34c/pkg/sql/opt_catalog.go#L638)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this `cat.Column` thing is something to pay attention to. 
</blockquote></details>

Yes, definitely! Unifying `catalog.Column` and `cat.Column` is a logical next step. I'll update the parent issue accordingly.
___
*[pkg/sql/opt_exec_factory.go, line 1956 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-MSkMkehC5Q2beLbip55:-MSmYxeV4R94JXkCezhi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52705f226fadfcc26b8a534ed0852a9d28fc9935/pkg/sql/opt_exec_factory.go#L1956)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The name of this function doesn't clearly describe to me its behavior; a comment would be helpful. 
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descriptor.go, line 404 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-MSiSVCX887V8-bbctkW:-MSmYzEoAOHw-mBE8gVn:b8o0aaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/50373fa7baf8df4655a6e7780952811946eb61d9/pkg/sql/catalog/descriptor.go#L404)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Should we comment this in terms of methods rather than fields?
</blockquote></details>

We do, it's at the bottom of the comment block, which follows the same structure as in `catalog.Index.Ordinal`. I might revise all of them once we're done introducing these new interfaces.
___
*[pkg/sql/catalog/descriptor.go, line 411 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-MSiSi7OCuUVf-uNO64y:-MSmZDyH-rlVft_DBsQ1:bb2jq28) ([raw file](https://github.com/cockroachdb/cockroach/blob/50373fa7baf8df4655a6e7780952811946eb61d9/pkg/sql/catalog/descriptor.go#L411)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	// Public returns true iff the column is active, i.e. readable, in the table
> 	// descriptor.
> 	Public() bool
> 	// WriteAndDeleteOnly returns true iff the column is a mutation in the
> 	// delete-and-write-only state in the table descriptor.
> 	WriteAndDeleteOnly() bool
> 	// DeleteOnly returns true iff the column is a mutation in the delete-only
> 	// state in the table descriptor.
> 	DeleteOnly() bool
> 	// Adding returns true iff the column is an add mutation in the table
> 	// descriptor.
> 	Adding() bool
> 	// Dropped returns true iff the column is a drop mutation in the table
> 	// descriptor.
> 	Dropped() bool
> ```
Are we happy with these as booleans? I know we did this on indexes but I have mixed feelings about it vs. an enum. 

----

After having reviewed this a bit, I think these works out pretty well, :shrug: 
</blockquote></details>

Thanks but I just copied the prevailing pattern. I usually like boolean method names to begin with `Is` or `Has` anyway.
___
*[pkg/sql/catalog/tabledesc/table.go, line 466 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59789#-MSk-8lA1vxJEfVDQ01U:-MSmZOi03TMCwrLaWdPf:buu7flq) ([raw file](https://github.com/cockroachdb/cockroach/blob/50373fa7baf8df4655a6e7780952811946eb61d9/pkg/sql/catalog/tabledesc/table.go#L466)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Note that an error will be returned if any of the names do not exist. 
</blockquote></details>

Good point. I updated the comment.


<!-- Sent from Reviewable.io -->

postamar commented:
bors r+
